### DISCUSSION TITLE
[front] User 정보에 대한 API 요청사항

### DISCUSSION THREAD

**Discussion by kimminkyeu** [3DPong/transcendence#14] [General]
### 채팅방에서의 API 요청은 이후에 추가할 예정입니다.

- Key는 바뀌어도 되고, 데이터 형식만 참고해주세요.

1. 프로필 구하기 
   Ex) GET /profile/{userId}
```jsonc
{
 /* userId        : "?" */
    imgSrc        : "프로필이미지URL",
    name          : "프로필이름",
}

```
2. 전적통계 구하기
   Ex) GET /statistic/{userId}
```jsonc
{
    totalGame  : ?,
    totalWin   : ?,
    totalLose  : ?,
    recent10Games : [ // 최대 10개 배열. 
            {
                winner : {  // --> profile 과 동일.
                    imgSrc : "URL",
                    name   : "이름",
                },
                loser  : { // --> profile 과 동일.
                    imgSrc : "URL",
                    name   : "이름",
                }
            },
            {
                // ...
            },
        ]
}
```

3. 친구 차단/해제 (미정)
   Ex) POST /{나의userId}/ban/{차단할친구userId} 

4. 친구 리스트 가져오기
    Ex) GET /{나의userId}/friends
    - Array of users
```jsonc
[
    {
        profile : { // profile 요청과 동일.
            imgSrc : "URL",
            name   : "이름",
        },
        //isBlocked     : [ true / false ],  // 차단 여부
        user_status : [ online/offline/...etc --> enum / ] 
    },
    {
        profile : { // profile 요청과 동일.
            imgSrc : "www...",
            name   : "schoe",
        },
       // isBlocked   : [ true/false ],              
        user_status : [ online/offline/...etc --> enum / ] 
    },
]
```

**Comment by JuneParkCode** [comment]
# USER CRUD
## CREATE
- create session 상태에서만 사용 가능 (이메일(인트라 식별자) 부여는 세션에서 관리됨)
### REQUEST
- `POST /user`
```jsonc
{
  "nickname" : string,
  "profile_url": url
}
```
### RESPONSE
`201 CREATED`
```jsonc
  {
  "user_id": number;
  "nickname": string;
  "profile_url": string;
  "wins": number;
  "losses": number;
  "total:" number;
  "level": number;
}
```
## READ
- 로그인된 세션에 대해서만 접근 가능
### REQUEST
- `GET /user/1`
### RESPONSE
```jsonc
{
  "user_id": number;
  "nickname": string;
  "profile_url": string;
  "wins": number;
  "losses": number;
  "total:" number;
  "level": number;
}
```

## UPDATE
- 로그인된 세션에 대해서만 접근 가능
- 본인 말고는 변경 불가능함
### REQUEST
```jsonc
{
  "nickname" ?:  string,
  "profile_url" ?: url,
  "two_factor" ?: boolean
}
```
### RESPONSE
```jsonc
{
  "user_id": number;
  "nickname": string;
  "profile_url": string;
  "wins": number;
  "losses": number;
  "total:" number;
  "level": number;
}
```

## DELETE
- 구현 됐는데 안쓴다고 생각하십쇼
### REQUEST
`DELETE /user/:userid`
### RESPONSE
`200`

**Reply by kimminkyeu** [reply]
login -> 42인증 

**1) 로그인 성공**  :---> 유저 정보 (userId만) 날라옴
-> 홈으로
```jsonc
200
{
  "status" : "SUCCESS" |  "SIGNUP_MODE", // 이 경우 회원가입모드로 ㄱ ㄱ
  user_id ?: 1  // SUCCESS일 경우일 때만 날라옴
}
```

**2) 회원가입모드**
- 회원가입 (프로필 설정) 페이지로
- **201 (유저 정보가 날라옴)** --> (회원가입 완료) --> 유저 정보 (userId만) 날라옴
      - response
 ```jsonc
201
{
 user_id : 1
}
 ```
- 홈으로

**Comment by kimminkyeu** [comment]
GET user/relation --> 유저간 관계 리스트 로드
- 쿼리 없음 ->  모든 관계
- ?relation=friend -> 친구 목록
- ?relatrion=block -> 블록된 유저 목록



--------
PUT user/relation  --> 유저간 상태 변경 ==> 내가 뭘 바꿀지 (block여부/친구삭제 등등) 이걸 같이 보내줘야함.

request
```jsonc
{
  userid : number, // 상대방 아이디
  friend:? boolean,
  block?: boolean,
}
```

response
```jsonc
{
  userid : number, // 상대방 아이디
  friend: boolean,
  block: boolean,
}
```

**Comment by kimminkyeu** [comment]
세션
- 서버에서 42 로그인 이후 세션을 부여를 해줍니다.
- 이는 cookie : connect_sid 에 그 값이 저장됨
- 세션이 정상이 아닐떄 401이 옵니다.
------------
400 Bad request : 대부분 dto. API call이 잘못된 거임. 
401은 무조건 세션비전상.
409는 로그인 후 다른 창에서 또 접속할 경우 --> 내가 "이미 접속된 유저입니다." 라고 띄워줘야함
